### ISSUE TITLE
Explore re-architecture of server-client interaction

### ISSUE THREAD

**Issue by mrecachinas** [LGSInnovations/jupyter-sigplot#13]
Per @sterre's recommendation:

Explore generically dispatching server-side calls to the client as much as possible; for example, a Python call to `sp.change_settings({'xmin': 0})` would be sent to the browser side as something like a json `{"function": "change_settings", "arguments": { "xmin": 0 }`  (ideally without special Python-side logic for the function name, either). It might be able to replace `oldArrays`, `oldHrefs`, and `inputs`, and possibly make future API changes to the Javascript SigPlot object transparently available to to widget.

This would require some pre-processing for those cases where we're currently moving files into a space accessible to the notebook server--this needs to be done generically.

**Comment by mrecachinas** [comment]
Available functions for use:
- `change_settings`
- `overlay_array` - will need to convert everything to Python lists
- `overlay_href` - this will incur a CORS, so the server will need to send an HTTP request for the file.
- `overlay_pipe` - does this now take an XMIDAS pipe since we're in Python? probably not.
- `overlay_websocket` - this will incur a CORS, so the server will likely need to setup a websocket connection to the websocket URI
- `push`
- `reload`

**Comment by sterre** [comment]
This would also help preserve layer ordering. At present, the client plots all arrays, then all hrefs, regardless of the order in which `overlay_array` and `overlay_href` was called. (Of course, there are other solutions to that particular problem.)

**Comment by mrecachinas** [comment]
To capture a bit more from Slack:

One idea is to send commands and args from server to client and execute the command with the corresponding args. The most recent idea that has been proposed is to whitelist specific plot commands on the server. This would look something like (in pseudo-code):

```python
class SigPlot(widgets.DOMWidget):
    def __getattr__(self, attr):
        if hasattr(self.child, attr)::
            def wrapper(*args, **kwargs):
                return getattr(self.child, attr)(*args, **kwargs)
        elif command in whitelisted_commands:
            def wrapper(*args, **kwargs):
                arguments = (args, kwargs)
                self.send_command(command, arguments)
        else:
            raise AttributeError(attr)
        return wrapper
```

**Comment by sterre** [comment]
This whitelist idea seems to strike a nice balance between ease of implementation and maintenance, and avoiding hard, weird, or nonsensical expressions.

I recognize the sketch above is just that, a sketch, but wanted to mention that it's helpful in production, when creating wrappers like that, to set the docstring for the wrapper based on the docstring for what it wraps. This makes `help()` and exception tracebacks much more legible.

**Comment by mrecachinas** [comment]
Implemented in 1d44f0d.
